core/src/main/java/org/infinispan/persistence/support/SegmentedAdvancedLoadWriteStoreAdapter.java

```diff
@@ -0,0 +1,205 @@
+package org.infinispan.persistence.support;
+
+import java.util.Objects;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Executor;
+import java.util.function.Predicate;
+
+import org.infinispan.commons.util.IntSet;
+import org.infinispan.distribution.ch.KeyPartitioner;
+import org.infinispan.persistence.internal.PersistenceUtil;
+import org.infinispan.persistence.spi.AdvancedCacheLoader;
+import org.infinispan.persistence.spi.CacheLoader;
+import org.infinispan.persistence.spi.CacheWriter;
+import org.infinispan.persistence.spi.ExternalStore;
+import org.infinispan.persistence.spi.InitializationContext;
+import org.infinispan.persistence.spi.MarshallableEntry;
+import org.infinispan.persistence.spi.SegmentedAdvancedLoadWriteStore;
+import org.infinispan.util.concurrent.CompletionStages;
+import org.reactivestreams.Publisher;
+
+import io.reactivex.rxjava3.core.Flowable;
+
+public class SegmentedAdvancedLoadWriteStoreAdapter<K, V> implements SegmentedAdvancedLoadWriteStore<K, V> {
+   private final CacheLoader<K, V> cacheLoader;
+   private final CacheWriter<K, V> cacheWriter;
+
+   private KeyPartitioner keyPartitioner;
+
+   public SegmentedAdvancedLoadWriteStoreAdapter(CacheLoader<K, V> cacheLoader) {
+      this.cacheLoader = Objects.requireNonNull(cacheLoader);
+      this.cacheWriter = cacheLoader instanceof CacheWriter ? (CacheWriter) cacheLoader : null;
+   }
+
+   public SegmentedAdvancedLoadWriteStoreAdapter(CacheWriter<K, V> cacheWriter) {
+      this.cacheWriter = Objects.requireNonNull(cacheWriter);
+      this.cacheLoader = cacheWriter instanceof CacheLoader ? (CacheLoader) cacheWriter : null;
+   }
+
+   @Override
+   public void init(InitializationContext ctx) {
+      keyPartitioner = ctx.getKeyPartitioner();
+      if (cacheLoader != null) {
+         cacheLoader.init(ctx);
+      } else {
+         cacheWriter.init(ctx);
+      }
+   }
+
+   @Override
+   public void start() {
+      if (cacheLoader != null) {
+         cacheLoader.start();
+      } else {
+         cacheWriter.start();
+      }
+   }
+
+   @Override
+   public void stop() {
+      if (cacheLoader != null) {
+         cacheLoader.start();
```

Comment:
`stop()` invokes `start()`?

Reply:
I didn't clean this part of the PR up yet. This class is actually to be deleted :)

Reply:
It was an initial attempt which has been superseded by NonBlockingStoreAdapter